src/backend/crossterm.rs

```diff
@@ -24,36 +23,310 @@ use crate::{
+    /// See [`crate::backend#alternate-screen`] for more information.
+    pub fn enter_alternate_screen(&mut self) -> io::Result<()> {
+        execute!(self.buffer, EnterAlternateScreen)?;
+        self.alternate_screen_entered = true;
+        Ok(())
+    }
+
+    /// Leaves the alternate screen.
+    ///
+    /// The backend will automatically leave the alternate screen when it is dropped.
+    ///
+    /// See [`crate::backend#alternate-screen`] for more information.
+    pub fn leave_alternate_screen(&mut self) -> io::Result<()> {
+        execute!(self.buffer, LeaveAlternateScreen)?;
+        self.alternate_screen_entered = false;
+        Ok(())
+    }
+
+    /// Enables mouse capture.
+    ///
+    /// The backend will automatically disable mouse capture when it is dropped.
+    ///
+    /// See [`crate::backend#mouse-capture`] for more information.
+    pub fn enable_mouse_capture(&mut self) -> io::Result<()> {
+        execute!(self.buffer, EnableMouseCapture)?;
+        self.mouse_capture_enabled = true;
+        Ok(())
+    }
+
+    /// Disables mouse capture.
+    ///
+    /// The backend will automatically disable mouse capture when it is dropped.
+    ///
+    /// See [`crate::backend#mouse-capture`] for more information.
+    pub fn disable_mouse_capture(&mut self) -> io::Result<()> {
+        execute!(self.buffer, DisableMouseCapture)?;
+        self.mouse_capture_enabled = false;
+        Ok(())
+    }
+}
+
+impl CrosstermBackend<Stderr> {
+    /// Creates a new `CrosstermBackend` using [`std::io::stderr()`] as the buffer.
+    ///
+    /// This method also enables [raw mode] and enters the [alternate screen].
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use ratatui::backend::{Backend, CrosstermBackend};
+    /// let mut backend = CrosstermBackend::on_stderr()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    ///
+    /// [raw mode]: crate::backend#raw-mode
+    /// [alternate screen]: crate::backend#alternate-screen
+    pub fn on_stderr() -> Result<CrosstermBackend<Stderr>, io::Error> {
+        CrosstermBackend::new(io::stderr())
+            .with_raw_mode()?
+            .with_alternate_screen()
+    }
+}
+
+impl CrosstermBackend<Stdout> {
+    /// Creates a new `CrosstermBackend` using [`std::io::stdout()`] as the buffer.
+    ///
+    /// This method also enables [raw mode] and enters the [alternate screen].
+    ///
+    /// # Example
+    ///
+    /// ```rust,no_run
+    /// # use ratatui::backend::{Backend, CrosstermBackend};
+    /// let mut backend = CrosstermBackend::on_stdout()?;
+    /// # std::io::Result::Ok(())
+    /// ```
+    ///
+    /// [raw mode]: crate::backend#raw-mode
+    /// [alternate screen]: crate::backend#alternate-screen
+    pub fn on_stdout() -> Result<CrosstermBackend<Stdout>, io::Error> {
+        CrosstermBackend::new(io::stdout())
+            .with_raw_mode()?
+            .with_alternate_screen()
+    }
+}
+
+impl Default for CrosstermBackend<Stderr> {
```

Comment:
I would assume that `Terminal::default()` would be a default that I can then use with no further setup for the common use case. Since that doesn't seem to be the case, I think 1 we should go with 1 here.

Reply:
It sort of does, raw mode and alternate screen are not strictly necessary to make an app work, they're mainly nice to have.

Reply:
I'm a little torn on whether `on_stderr()` should do `raw_mode` and `alternate_screen` as well. 

Maybe it's just a naming thing. I do like having a convenience function that does `stderr + raw_mode + alternate_screen`, but if I were reading code that used this for the first time, I'd be very surprised to learn that a function called `on_stderr()` also activated `raw_mode` and also toggled to the `alternate_screen`.

What are your thoughts on this? Can we brainstorm alternative names for this?

Reply:
Maybe `on_stderr_tui()`?

Reply:
> Maybe `on_stderr_tui()`?

I'm not sure that really communicates anything about defaults better.

Happy to brainstorm some more ideas on this if you'd like. I'd start with what it is that we want to configure and then design something that handles the normal use cases pretty seamlessly:

Options:
- Backend: Crossterm / Termion / Termwiz
- Buffer choice: Stdout / Stderr / (any impl Write) / no choice for Termwiz right now
- Raw mode: yes / no. (Termion requires this to construct, Termwiz always enables this)
- Alternate Screen: yes / no. (Termion is optional, but changes the buffer type, Termwiz always enables this)
- Terminal Viewport type (Inline, Fixed, Fullscreen)

The two most common use cases (due to compatibility and the types of apps people write) are:
- Full screen Crossterm on stdout with raw mode and alternate screen
- Full screen Crossterm on stderr with raw mode and alternate screen

I really wanted to find something that does the most useful thing in one call here. Perhaps a proper builder approach would be better - that way we can defer the failure check to `build()` and the actual `default()` call never fails. E.g.:

```rust
let backend = CrosstermBackendBuilder::<Stdout>::default().build()?
```
I don't like that because the generic makes it unweildy

Perhaps:
```rust
let terminal = TerminalBuilder::with_crossterm().on_stderr().build()?
```

Using this approach the choices that we make in the defaults are part of the builder and can be communicated as such, not part of the backend / terminal types. Not sure if this is better though and it's different from all the other builders in Ratatui (which is the reason I didn't give it much consideration).

Reply:
> I really wanted to find something that does the most useful thing in one call here. 

I really like the functionality. I'm just wondering if there's a better way to communicate what is going on there.

> ```rust
> let terminal = TerminalBuilder::with_crossterm().on_stderr().build()?
> ```

Oddly enough, I actually like it.

I think where my comment is coming from is that `on_$io` doesn't seem to convey enough. `build()` tells me something is being "built" that needs to be "destroyed" which maps with switching to alternate screen and back. 

Just tossing some other ideas out there:

```rust
// Option 1:
let terminal = TerminalBuilder::with_crossterm().build(stderr=true)?
// Option 2:
let terminal = TerminalBuilder::with_crossterm().init(stdout=false)?
// Option 3:
let terminal = TerminalBuilder::with_crossterm().setup()? // defaults to stderr. If people want stdout they can build the old fashion way. 
// Option 4:
let terminal = TerminalBuilder::with_crossterm().tui_on_stderr()?
let terminal = TerminalBuilder::with_crossterm().tui_on_stdout()?
```

Reply:
> let terminal = TerminalBuilder::with_crossterm().on_stderr().build()?

> let terminal = TerminalBuilder::with_crossterm().setup()?

These options look the most sane to me.

Reply:
Of those two I prefer

>     let terminal = TerminalBuilder::with_crossterm().on_stderr().build()?

Because setup()? seems weird and like it shouldn't return anything.

Reply:
But I think it is common in rust to have builder methods with names like `setup()` that return.

Here's code from `tracing_subscriber`:

```rust
  let file_subscriber = tracing_subscriber::fmt::layer()
    .with_file(true)
    .with_line_number(true)
    .with_writer(log_file)
    .with_target(false)
    .with_ansi(false)
    .with_filter(tracing_subscriber::filter::EnvFilter::from_default_env());
```

The `config` crate has examples like this:

```rust
    let settings = Config::builder()
        // File::with_name(..) is shorthand for File::from(Path::new(..))
        .add_source(File::with_name("examples/glob/conf/00-default.toml"))
        .add_source(File::from(Path::new("examples/glob/conf/05-some.yml")))
        .add_source(File::from(Path::new("examples/glob/conf/99-extra.json")))
        .build()
        .unwrap();
```

`notcurses` has:

```
impl NotcursesBuilder {
    /// Returns a new default `NotcursesBuilder`, without banners.
    pub fn new() -> Self {
        Self::default()
    }

    /// Returns a `Notcurses` instance.
    pub fn build(self) -> Result<Notcurses> {
        Notcurses::lock_notcurses()?;
        let nc = unsafe { Nc::with_options(self.options.build())? };
        Ok(Notcurses {
            nc,
            options: self.options,
        })
    }
}
```

And [all the methods in `NotcursesBuilder`](https://github.com/dankamongmen/notcurses-rs/blob/4caf961306866ddc70f62d52be9083779b48d8f9/src/notcurses/builder.rs#L49) seem to be verbs.

What do you think about prefixing everything with `with`? e.g.:

```rust
let terminal = TerminalBuilder::with_crossterm().with_stderr().build()?
```